<h1><center>Laboratorio 9: Benchmark Estadístico con Reddit 🧮</center></h1>

<center><strong>MDS7202: Laboratorio de Programación Científica para Ciencia de Datos</strong></center>

### Cuerpo Docente:

- Profesor: Pablo Badilla
- Auxiliar: Ignacio Meza D.
- Ayudante: Constanza Peña

### Equipo: SUPER IMPORTANTE - notebooks sin nombre no serán revisados

- Nombre de alumno 1: Santiago Valdivieso
- Nombre de alumno 2:


### **Link de repositorio de GitHub:** `http://....`

# Temas a tratar

- Optimización de Código en Python.
- Utilización de librerías para medir el tiempo de ejecución de funciones.
- Métodos para optimizar el rendimiento de las funciones.

# Reglas:

- Fecha de entrega: 3/12/2021
- **Grupos de 2 personas**
- **Ausentes** deberán realizar la actividad solos. 
- Cualquier duda fuera del horario de clases al foro. Mensajes al equipo docente serán respondidos por este medio.
- Prohibidas las copias. 
- Pueden usar cualquer matrial del curso que estimen conveniente.

# Objetivos principales del laboratorio

- Obtener datos desde Reddit a partir de la librería `praw` y visualizar cuales post son más probables que sean puntuados positivamente.
- Aplicar un atajo estadístico para obtener la mean posterior de datos.
- Optimizar a través de librerías funciones.

#Importamos librerias utiles 😸

In [4]:
%%capture
!pip install praw
!pip install line_profiler
!pip install numba

%load_ext autoreload
%autoreload 2
from sklearn.cluster import KMeans
import sys
import praw
import time
import numpy as np
import pandas as pd
from numba import jit
import plotly.express as px
from functools import lru_cache
from IPython.core.display import Image

# 1. Recomendando Posts de Subreddit

<p align="center">
  <img src="https://styles.redditmedia.com/t5_2rer8/styles/bannerBackgroundImage_xzrb6my122181.jpg">
</p>

Unos intrépidos alumnos del curso, quienes frecuentemente participan en subreddits y foros de reviews, se preguntan lo siguiente: ¿Podremos confiar que un post es bueno, si este tiene solamente 1 o 3 votos positivos?. los compañeros, creen que esto claramente no representa una opinión general, ya que estamos mucho menos seguros acerca de la verdadera proporción de votos a favor de los comentarios con pocos datos. ¿Pero cómo podemos obtener una representación más creíble para este problema?.

Lo señalado forma parte de un problema estadísticos, donde a través del cálculo de la posterior se puede conocer que tan probable es que un post sea bueno. Para efectos de este laboratorio, no se exige un conocimiento previo para resolver este problema, simplemente se deberá aplicar las ecuaciones presentadas más adelante (De igual forma si quedan interesados sobre el tema se les invita a tomar el ramo [CC6104](https://github.com/dccuchile/CC6104)).

## 1.1 Obtención de Subrredits y Análisis [1 Punto]

Para estudiar que tan probable es que un post sea positivo se comenzará cargando datos reales del subreddit chile desde reddit (si usted desea puede cambiar el subreddit a uno de su gusto). Para esto le proponemos la utilización de la función que aparece mas abajo, la que presenta un usuario ya creado por el equipo docente. Dese un tiempo para entender que hace cada parte de la función, visualizando que se obtiene de estas.

Revisada la función, utilice un **perfilador** para monitorear el tiempo y memoria que les toma a cada liena de código para ser ejecutada. Señale cuales son los procesos que mas tiempo consumen en la ejecución del Código, comentando si es posible mejorar el desempeño de la función.

**TO-DO:**
- [ ] Estudiar la función propuesta por el equipo docente.
- [ ] Estudiar los tiempos de ejecución del código a través de un perfilador.
- [ ] Estudiar la memoria ocupada por el código a través de un perfilador.
- [ ] Comente los resultados. ¿Qué lineas ocupan más memoria o más/menos tiempo?¿Por qué?

In [2]:
%load_ext line_profiler

In [5]:
def praw_reddit(nombre_subreddit="chile", n_hot=1000):
    reddit = praw.Reddit(
        client_id="-w2hyFINxZ8T3g",
        client_secret="zGPCI4s3g6Ic6AsRi7vIpP0NoxbFdw",
        password="ClasesMDS7202",
        user_agent="Clases",
        username="DocenciaDataScience",
        check_for_async=False,
    )
    subreddit = reddit.subreddit(nombre_subreddit)

    votes, post, url = {}, {}, {}
    top_submissions = list(subreddit.hot(limit=n_hot))
    for it, top_n in enumerate(range(50, len(top_submissions), 50)):
        top_n_submissions = top_submissions[:top_n]
        upvotes, downvotes, url[it], post[it] = [], [], [], []

        for submission in top_n_submissions:
            try:
                ratio = submission.upvote_ratio
                ups = int(
                    round((ratio * submission.score) / (2 * ratio - 1))
                    if ratio != 0.5
                    else round(submission.score / 2)
                )
                upvotes.append(ups)
                downvotes.append(ups - submission.score)
                post[it].append(submission.title)
                url[it].append(submission.url)
            except Exception as e:
                continue
        votes[it] = np.array([upvotes, downvotes]).T
    return votes, post, url


In [6]:
votes, post, url = praw_reddit()

In [60]:
%lprun -f praw_reddit praw_reddit()

In [57]:
%%file bench_module.py
import numpy as np
import praw
def praw_reddit(nombre_subreddit="chile", n_hot=1000):
    reddit = praw.Reddit(
        client_id="-w2hyFINxZ8T3g",
        client_secret="zGPCI4s3g6Ic6AsRi7vIpP0NoxbFdw",
        password="ClasesMDS7202",
        user_agent="Clases",
        username="DocenciaDataScience",
        check_for_async=False,
    )
    subreddit = reddit.subreddit(nombre_subreddit)

    votes, post, url = {}, {}, {}
    top_submissions = list(subreddit.hot(limit=n_hot))
    for it, top_n in enumerate(range(50, len(top_submissions), 50)):
        top_n_submissions = top_submissions[:top_n]
        upvotes, downvotes, url[it], post[it] = [], [], [], []

        for submission in top_n_submissions:
            try:
                ratio = submission.upvote_ratio
                ups = int(
                    round((ratio * submission.score) / (2 * ratio - 1))
                    if ratio != 0.5
                    else round(submission.score / 2)
                )
                upvotes.append(ups)
                downvotes.append(ups - submission.score)
                post[it].append(submission.title)
                url[it].append(submission.url)
            except Exception as e:
                continue
        votes[it] = np.array([upvotes, downvotes]).T
    return votes, post, url

Overwriting bench_module.py


In [61]:
from bench_module import praw_reddit
%mprun -f praw_reddit praw_reddit()

**Respuesta:**

La línea que ocupa más tiempo de procesamiento de la función praw_reddit es la 13 "top_submissions = list(subreddit.hot(limit=n_hot))" la cual tarda poco más de 13 segundos, esto ocurre porque debe extraer 1000 submisions de reddit conectándose a la red. Esta es la que toma más tiempo del código 99,3%.

Por otro lado, las líneas que tardan menos tiemmpo son de la 3 a la 7 y la 33 que es el return, que demoran 1 micro segundo o menos, que son las variables con las cuales se entra a reddit, por lo que representan sólo una definición de una variable para el procesamiento y el return que simplemente devuelve el valor que ya se tiene guardado, por lo que es exigente en rendimiento.

Respecto a la memoria la línea que más ocupa es la primera de la función, "def praw_reddit(nombre_subreddit="chile", n_hot=1000):" la cual toma 406,2 Mib, en usarla, sin embargo esta no aumenta en las siguientes líneas porque los numeros  de ocurrencia son valores muy bajos(ninguno mayor a 10000)



## 1.2 LRU y Análisis de Tiempo con Cache [1 Punto]

Respondan las siguientes preguntas:

1.	¿Qué es la memoria cache y a que se refiere las siglas LRU?
2.	¿Cuáles son los costos que tiene la aplicación de técnicas de Caching?
3. ¿Cuál es la consecuencía de ocupar caching en la función anterior?.

Respondidas las preguntas, se le solicita que aplique alguna técnica de caching para mejorar el desempeño de la función `praw_reddit`. Para esto compare solo el tiempo de ejecución del algoritmo con y sin caching, señalando el tiempo total de ejecución y el tiempo promedio que le toma ejecutar cada loop a la función. Con lo anterior, ¿es posible visualizar mejoras en este caso?.


**TO-DO:**
- [ ] Responder las preguntas.
- [ ] Mejorar el código con cache.
- [ ] Explicar las consecuencias de usar lru_cache sobre praw_reddit

**Respuestas Teóricas:**

`Escriba su respuesta aquí`


¿Cuál es la consecuencía de ocupar caching en la función anterior?.

la memoria cache es un espacio que puede estar ubicado en memoria (RAM), disco o almacenada de manera remota. El acto de guardar resultados en memoria para luego utilizarlos de manera directa se denomina memorización y es una forma de chaching para mejorar el rendimiento de aplicaciones por medio de un uso eficiente de la memoria, una de las ideas tras esta premisa es la de guardar los resultados de operaciones intensivas. Las siglas LRU representan e l algoritmo Least Recently Used, el cual descarta primero los elementos menos usados recientemente. El algoritmo lleva el seguimiento de lo que se va usando, lo que resulta caro si se quiere hacer con precisión. La implementación de esta técnica requiere llevar la cuenta de la edad de cada elemento de caché y buscar el menos usado sobre la base de ella.

Las técnicas de caching tienen un costo, este radica en aumentar el consumo de memoria, si esta memoria esta localizada en disco, el acceso puede ser muy lento y el rendimiento puede decaer drásticamente.

In [62]:
from functools import lru_cache

In [68]:
@lru_cache(maxsize=1000)
def praw_reddit(nombre_subreddit="chile", n_hot=1000):
    reddit = praw.Reddit(
        client_id="-w2hyFINxZ8T3g",
        client_secret="zGPCI4s3g6Ic6AsRi7vIpP0NoxbFdw",
        password="ClasesMDS7202",
        user_agent="Clases",
        username="DocenciaDataScience",
        check_for_async=False,
    )
    subreddit = reddit.subreddit(nombre_subreddit)

    votes, post, url = {}, {}, {}
    top_submissions = list(subreddit.hot(limit=n_hot))
    for it, top_n in enumerate(range(50, len(top_submissions), 50)):
        top_n_submissions = top_submissions[:top_n]
        upvotes, downvotes, url[it], post[it] = [], [], [], []

        for submission in top_n_submissions:
            try:
                ratio = submission.upvote_ratio
                ups = int(
                    round((ratio * submission.score) / (2 * ratio - 1))
                    if ratio != 0.5
                    else round(submission.score / 2)
                )
                upvotes.append(ups)
                downvotes.append(ups - submission.score)
                post[it].append(submission.title)
                url[it].append(submission.url)
            except Exception as e:
                continue
        votes[it] = np.array([upvotes, downvotes]).T
    return votes, post, url


In [69]:
%%timeit
votes, post, url = praw_reddit()

The slowest run took 63309883.73 times longer than the fastest. This could mean that an intermediate result is being cached.
1 loop, best of 5: 223 ns per loop


**Respuesta de lru_cache sobre praw_reddit**
Podemos ver que al aplicar %%timeit que el tiempo que demoró la iteración más lenta de la función es 63309883.73 veces mayor que la más rápida. esto es porque el resultado de la función está guardada en el cache. En este caso en particular debido a que se van generando datos por distintos tamaños de batches, se repiten las funciones distintas veces, lo que hace que al guardarlas en el caché el rendimiento total sea mucho mayor

## 1.3 Obtención de Mean Posterior y Standard Error [1 Punto]

Una forma de obtener la mean posterior y el Standard Error de los datos de reddit es aplicando un atajo de inferencia estadística. Este atajo se define de la siguiente forma:

Sea **u** los votos positivos y **d** los votos negativos del subreddit, tendremos que:

$$a = 1+u$$

$$b = 1+d$$

$$\sigma= 1.65\sqrt(\dfrac{ab}{(a + b)^2(a + b + 1)})$$

$$\mu = \dfrac{a}{a+b}$$

Donde $\mu$ es la mean posterior y $\sigma$ el standard error.

Con lo anterior, genere dos funciones que tengan como salida $\mu$ y $\sigma$ de acuerdo a las ecuaciones señaladas. La primera función, deberá ser construida sin el uso de numpy, aplicando for y aplicando comandos nativos de Python. Por otro lado, deberá generar una segunda función con el uso exclusivo de numpy. **OJO** que las funciones deben tener como entrada solo un elemento del diccionario votes (por ejemplo `votes[1]`), por lo que estas no deben tener como entrada el conjunto completo de datos.

In [143]:
def funcion11(votos):
  Media=[]
  desv=[]
  for i in range(len(votos)):
    positivos = votos[i][0]
    negativos = votos[i][1]
    a=1+positivos
    b=1+negativos
    media=a/(a+b)
    desviación=1.65*(((a*b)/(((a*b)**2)*(a+b+1)))*0.5)
    Media.append(media)
    desv.append(desviación)

  return Media,desv

In [139]:
def funcion22(votos):
  a=1+votos[:,0]
  b=1+votos[:,1]
  media=a/(a+b)
  desviación=1.65*(((a*b)/(((a*b)**2)*(a+b+1)))*0.5)
  return media, desviación

In [132]:
%timeit funcion11(votes[0])

10000 loops, best of 5: 165 µs per loop


In [140]:
%timeit funcion22(votes[0])

The slowest run took 8.70 times longer than the fastest. This could mean that an intermediate result is being cached.
100000 loops, best of 5: 13.4 µs per loop


In [150]:
funcion11(votes[0])

([0.8130841121495327,
  0.928395061728395,
  0.9274193548387096,
  0.8379310344827586,
  0.979381443298969,
  0.8909090909090909,
  0.9340659340659341,
  0.9259259259259259,
  0.9444444444444444,
  0.9555555555555556,
  0.9125,
  0.8269230769230769,
  0.8913043478260869,
  0.8472222222222222,
  0.8611111111111112,
  0.6746987951807228,
  0.9384288747346072,
  0.8846153846153846,
  0.7777777777777778,
  0.8846153846153846,
  0.7674418604651163,
  0.9230769230769231,
  0.8131868131868132,
  0.9788235294117648,
  0.9274809160305344,
  0.9757785467128027,
  0.8777777777777778,
  0.9067164179104478,
  0.6666666666666666,
  0.8888888888888888,
  0.9523809523809523,
  0.8095238095238095,
  0.9795158286778398,
  0.918918918918919,
  0.9375,
  0.8939393939393939,
  0.8675213675213675,
  0.9130434782608695,
  0.9714285714285714,
  0.863013698630137,
  0.7142857142857143,
  0.926605504587156,
  0.9230769230769231,
  0.9953917050691244,
  0.7777777777777778,
  0.8152173913043478,
  0.7142857142857

In [151]:
funcion22(votes[0])

(array([0.81308411, 0.92839506, 0.92741935, 0.83793103, 0.97938144,
        0.89090909, 0.93406593, 0.92592593, 0.94444444, 0.95555556,
        0.9125    , 0.82692308, 0.89130435, 0.84722222, 0.86111111,
        0.6746988 , 0.93842887, 0.88461538, 0.77777778, 0.88461538,
        0.76744186, 0.92307692, 0.81318681, 0.97882353, 0.92748092,
        0.97577855, 0.87777778, 0.90671642, 0.66666667, 0.88888889,
        0.95238095, 0.80952381, 0.97951583, 0.91891892, 0.9375    ,
        0.89393939, 0.86752137, 0.91304348, 0.97142857, 0.8630137 ,
        0.71428571, 0.9266055 , 0.92307692, 0.99539171, 0.77777778,
        0.81521739, 0.71428571, 0.97350993, 0.88372093, 0.9379845 ]),
 array([4.39016603e-06, 1.86355439e-07, 2.37444578e-07, 2.48231464e-07,
        4.43071966e-05, 6.32009561e-06, 1.75831202e-05, 2.23577236e-05,
        1.22194163e-06, 2.63544595e-05, 1.99318692e-05, 4.02223197e-05,
        8.56253243e-05, 1.68425780e-05, 1.43853531e-04, 6.49565382e-06,
        1.36361473e-07, 4.4283

## 1.4 Comparación de Rendimiento [2 Puntos]

Creadas las funciones de la sección 1.3, de forma iterativa compare los diferentes batches de datos que contiene el diccionario `votes`. Para esto genere un gráfico utilizando plotly, donde se pueda observar las diferencias de tiempo que toma ejecutar las diferentes cantidades de datos. ¿es posible observar una diferencia? ¿a qué se deberá esto?.

Aplique el compilador **Numba** sobre las funciones 1.3 y compare el tiempo de ejecución con los obtenidos sin el compilador. 

In [144]:
from numba import jit
@jit(nopython=True)
def funcion11_numba(votos):
  positivos=0
  negativos=0
  for i in range(len(votos)):
    positivos = positivos+ votos[i][0]
    negativos = negativos + votos[i][1]
  a=1+positivos
  b=1+negativos
  media=a/(a+b)
  desviación=1.65*(((a*b)/(((a*b)**2)*(a+b+1)))*0.5)
  return media, desviación

@jit(nopython=True)
def funcion22_numba(votos):
  a=1+np.sum((votos)[:,0])
  b=1+np.sum((votos)[:,1])
  media=a/(a+b)
  desviación=1.65*(((a*b)/(((a*b)**2)*(a+b+1)))*0.5)
  return media, desviación

In [148]:
tiempo2=[]
for i in range(len(votes)):
  t0 = time.process_time()
  funcion11(votes[i])
  t1 = time.process_time()
  funcion11_numba(votes[i])
  t2 = time.process_time()
  funcion22(votes[i])
  t3 = time.process_time()
  a=np.array(votes[i])
  t4 = time.process_time()  
  funcion22_numba(a)
  t5 = time.process_time()
  tiempo2.append([len(votes[i]),(t1-t0),(t2-t1),(t3-t2),(t5-t4)])

df = pd.DataFrame(np.array(tiempo2), 
             columns=['Cantidades', 
                      'Función Nativa',
                      'Función Nativa Numba', 
                      'Función Numpy',
                      'Función Numpy Numba'])

In [149]:
fig = px.line(df, x="Cantidades", y=['Función Nativa','Función Nativa Numba','Función Numpy','Función Numpy Numba'])
fig.show()

Ahora, grafiquen los tiempos tomados y comente los desempeños.

**Nota:** Ejecuten las funciones compiladas con algún ejemplo antes de graficar. De lo contrario, les indicará el tiempo de compilación, cosa que no queremos medir

Podemos ver que el orden de rendimiento de las funciones son las Numba tanto en numpy como nativa son las más rápidas. Luego la función cn nupy y finalmente la que más aumenta el tiempo de ejecución según la cantidad de elementos es la nativa.

Si se hac zoom en el gráfico (no se alcanza a apreciar en sin zoom) la función numpy usando numba es más rápida que la nativa.

## 1.5 Plot de Resultados [1 Punto]

Llego la hora de visualizar los resultados obtenidos, para esto solo ejecute las siguientes celdas y observe lo que sucede :3. ¿Qué logra observar de los resultados?, ¿La solución resulta trivial?

En esta sección esperamos que solo comenten con lo que logran visualizar de los dataframes.

In [152]:
votos = votes[len(votes) - 1]
posteo = post[len(post) - 1]
print("lower bounds aproximados:")
posterior_mean, std_err = funcion22(votos)
lb = posterior_mean - std_err
print("\nTop 20 post ordenador por el limite inferior:\n")
order = np.argsort(-lb)
vote_post = {
    'Votos (+)': votos[order[:20], 0],
    'Votos (-)': votos[order[:20], 1],
    'Post': np.array(posteo)[order[:20]],
    'url': np.array(url[len(url) - 1])[order[:20]]
}
df = pd.DataFrame(data=vote_post)
ordered_post = df.Post

r_order = order[::-1][-20:]
data_dic = {
    'mean': posterior_mean[r_order],
    'std_err': std_err[r_order],
    'post': ordered_post[::-1]
}
df = pd.DataFrame(data=data_dic)
fig = px.scatter(df, x="mean", y="post", error_x="std_err")
fig.show()

lower bounds aproximados:

Top 20 post ordenador por el limite inferior:



**Respuesta:**

En la visualización se puede apreciar los 20 post de reddit ordenados de mayor a menor por su media posterior, la que estadísticamente representa la probabilidad un post sea bueno. Podemos ver que el post con mayor probalididad es que le guste al lector es el que empieza con ¡Estamos de vuelrta!, con un valor de 0.998% de probabilidad de que al lector de lo considere positivo. En cambio el post número 20 dentro de este ranking es "El crossover que nadie pidió, pero que todos necesitabamos". 

Además podemo obtener las desviaciones estandar de cada uno de estos post, la con mayor desvuación es la de "Sernac acusa a Banco de Chile (...)". 

<p align="center">
  <img src="https://media1.tenor.com/images/fb5bf7cc5a4acb91b4177672886a88ba/tenor.gif?itemid=5591338">
</p>